server/api/micro_service.go

```diff
@@ -0,0 +1,91 @@
+package api
+
+import (
+	"fmt"
+	"net/http"
+
+	"github.com/gorilla/mux"
+	"github.com/pingcap/log"
+	"github.com/tikv/pd/pkg/mcs/discovery"
+	"github.com/tikv/pd/server"
+	"github.com/unrolled/render"
+	"go.uber.org/zap"
+)
+
+type microServiceHandler struct {
+	svr *server.Server
+	rd  *render.Render
+}
+
+func newMicroServiceHandlerHandler(svr *server.Server, rd *render.Render) *microServiceHandler {
+	return &microServiceHandler{
+		svr: svr,
+		rd:  rd,
+	}
+}
+
+// @Tags     members
+// @Summary  Get all members of the cluster for the specified service.
+// @Produce  json
+// @Success  200  {object}  []string
+// @Router   /ms/members/{service} [get]
+func (h *microServiceHandler) GetMembers(w http.ResponseWriter, r *http.Request) {
+	if !h.svr.IsAPIServiceMode() {
+		h.rd.JSON(w, http.StatusServiceUnavailable, "not support micro service")
+		return
+	}
+
+	if service := mux.Vars(r)["service"]; len(service) > 0 {
+		resps, err := discovery.GetMembers(service, h.svr.GetClient())
+		if err != nil {
+			h.rd.JSON(w, http.StatusInternalServerError, err.Error())
+			return
+		}
+		if resps == nil {
+			h.rd.JSON(w, http.StatusNotFound, fmt.Sprintf("no members for %s", service))
+			return
+		}
+
+		var apis []string
+		for _, resp := range resps.Responses {
+			for _, keyValue := range resp.GetResponseRange().GetKvs() {
+				var entry discovery.ServiceRegistryEntry
+				if err = entry.Deserialize(keyValue.Value); err != nil {
+					log.Info("deserialize failed", zap.String("key", string(keyValue.Key)), zap.Error(err))
+				}
+				apis = append(apis, entry.ServiceAddr)
+			}
+		}
+		h.rd.JSON(w, http.StatusOK, apis)
+		return
+	}
+
+	h.rd.JSON(w, http.StatusInternalServerError, "not support service")
+}
+
+// @Tags     leader
+// @Summary  Get the leader of the cluster for the specified service.
+// @Produce  json
+// @Success  200  {object}  pdpb.Member
+// @Router   /ms/leader/{service} [get]
+func (h *microServiceHandler) GetLeader(w http.ResponseWriter, r *http.Request) {
+	if !h.svr.IsAPIServiceMode() {
+		h.rd.JSON(w, http.StatusServiceUnavailable, "not support micro service")
+		return
+	}
+	if service := mux.Vars(r)["service"]; len(service) > 0 {
+		leader, _, err := discovery.GetMCSPrimary(service, r.URL.Query().Get("keyspace_id"), h.svr.GetClient())
```

Comment:
How about making keyspace id as an optional parameter?

Reply:
Currently, keyspace id is an optional para, which will be checked in `getMCSPrimaryPath`